src/base/intrusive_list.h

```diff
@@ -48,26 +48,56 @@ namespace perfetto::base {
 
 namespace internal {
 
+// ListNode is used both in (1) actual list nodes; (2) as the list head, to
+// point to the first and last element in the list.
+// When prev_/next_ of a node point to the list head, the address has the LSB
+// set. This has two purposes: (1) identify the list head to stop the iterator;
+// (2) prevent bugs which try to dereference the list head casting it into a T.
 struct ListNode {
-  ListNode* prev;
-  ListNode* next;
+  uintptr_t prev_;
+  uintptr_t next_;
 };
 
+// This function masks away the LSB retutning a pointer to a ListNode. This can
+// be used when we want to dereference a prev_/next_ pointer and we acknowledge
+// that we might be operating on the list head rather than a node.
+inline ListNode* MaybeHead(uintptr_t p) {
+  return reinterpret_cast<ListNode*>(p & ~uintptr_t(1));
+}
+
 // IntrusiveList's Base class to factor out type-independent code (avoid binary
 // bloat)
 class ListOps {
  public:
   void PushFront(ListNode* node);
+  void PushBack(ListNode* node);
   void PopFront();
+  void PopBack();
   void Erase(ListNode* node);
+  bool empty() const { return head_.next_ == head_masked(); }
 
-  ListNode* front_{nullptr};
-  size_t size_{0};
+  uintptr_t head_masked() const {
+    return reinterpret_cast<uintptr_t>(&head_) | 1;
+  }
+  ListNode head_{head_masked(), head_masked()};
```

Comment:
I think after reading the patch a bunch I understand this but I think calling this "head" is quite misleading. THis is more a "close the loop" node (i.e. prev points to tail, next points to head). I think calling it something different would help with readaiblity.

Reply:
I used the name list_head because that's really the same name that the kernel uses for precisely the same reason since the beginning of times:
https://elixir.bootlin.com/linux/2.1.45/source/include/linux/list.h

Reply:
Fair enough, I still think it's quite misleading but who am I to question the kernel :)

Reply:
renamed to head_and_tail